File: 403. Frog Jump.py
class Solution(object):
    def canCross(self, stones):
        """
        :type stones: List[int]
        :rtype: bool
        """
        n = len(stones)
        if stones[1] != 1:
            return False
        dp = {}
        for i in stones:
            dp[i] = []
        dp[1].append(1)
        for i in stones:
            for st in dp[i]:
                for k in [-1, 0, 1]:
                    x = i + st + k
                    if x in dp:
                        y = st + k
                        if y not in dp[x]:
                            dp[x].append(y)

        return len(dp[stones[-1]]) > 0
        

File: 403. Frog Jump.go
func canCross(stones []int) bool {
	n := len(stones)
	if stones[1] != 1 {
		return false
	}
	if n == 2 {
		return true
	}
	dp := make([][]bool, n)
	for i := range dp {
		dp[i] = make([]bool, n+1)
	}
	dp[1][1] = true
	for i := 2; i < n; i++ {
		for j := i - 1; j > 0; j-- {
			k := stones[i] - stones[j]
			if k > j+1 {
				break
			}
			dp[i][k] = dp[j][k-1] || dp[j][k] || dp[j][k+1]
			if i == n-1 && dp[i][k] {
				return true
			}
		}
	}

	return false
}